crates/valence_boss_bar/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+        for viewer in boss_bar_viewers.viewers.iter_mut() {
+            let mut client = clients.get_mut(*viewer).unwrap();
+            client.write_packet(&BossBarS2c {
+                id: id.0,
+                action: BossBarAction::UpdateStyle(style.color, style.division),
+            });
+        }
+    }
+}
+
+/// System that sends a bossbar update flags packet to all viewers of a boss bar
+/// that has had its flags updated.
+fn handle_boss_bar_flags_update(
+    mut boss_bars: Query<(&UniqueId, &BossBarFlags, &mut BossBarViewers), Changed<BossBarFlags>>,
+    mut clients: Query<&mut Client>,
+) {
+    for (id, flags, mut boss_bar_viewers) in boss_bars.iter_mut() {
+        for viewer in boss_bar_viewers.viewers.iter_mut() {
+            let mut client = clients.get_mut(*viewer).unwrap();
+            client.write_packet(&BossBarS2c {
+                id: id.0,
+                action: BossBarAction::UpdateFlags(*flags),
+            });
+        }
+    }
+}
+
+/// System that sends a bossbar add/remove packet to all viewers of a boss bar
+/// that just have been added/removed.
+fn handle_boss_bar_viewers_update(
+    mut boss_bars: Query<
+        (
+            &UniqueId,
+            &BossBarTitle,
+            &BossBarHealth,
+            &BossBarStyle,
+            &BossBarFlags,
+            &mut BossBarViewers,
+        ),
+        Changed<BossBarViewers>,
+    >,
+    mut clients: Query<&mut Client>,
+) {
+    for (id, title, health, style, flags, mut boss_bar_viewers) in boss_bars.iter_mut() {
+        let old_viewers = &boss_bar_viewers.old_viewers;
+        let current_viewers = &boss_bar_viewers.viewers;
+
+        let mut added_viewers = Vec::new();
+        let mut removed_viewers = Vec::new();
+
+        for viewer in current_viewers.iter() {
+            if !old_viewers.contains(viewer) {
+                added_viewers.push(*viewer);
+            }
+        }
+
+        for viewer in old_viewers.iter() {
+            if !current_viewers.contains(viewer) {
+                removed_viewers.push(*viewer);
+            }
+        }
+
+        for added_viewer in added_viewers {
+            let mut client = clients.get_mut(added_viewer).unwrap();
+            client.write_packet(&BossBarS2c {
+                id: id.0,
+                action: BossBarAction::Add {
+                    title: title.0.clone(),
+                    health: health.0,
+                    color: style.color,
+                    division: style.division,
+                    flags: *flags,
+                },
+            });
+        }
+
+        for removed_viewer in removed_viewers {
+            if let Ok(mut client) = clients.get_mut(removed_viewer) {
+                client.write_packet(&BossBarS2c {
+                    id: id.0,
+                    action: BossBarAction::Remove,
+                });
+            }
+        }
+
+        boss_bar_viewers.old_viewers = boss_bar_viewers.viewers.clone();
```

Comment:
There's some O(n²) work happening here that could be avoided with sets. Also the allocation of temporary `added_viewers` and `removed_viewers` vecs shouldn't be necessary.